### DISCUSSION TITLE
Following Github Actions tutorial but getting an npm EACCES error

### DISCUSSION THREAD

**Discussion by seddonr** [community/community#27033] [Actions]
<p>I’m following the tutorial here: <a href="https://docs.github.com/en/actions/learn-github-actions/introduction-to-github-actions#create-an-example-workflow" class="inline-onebox" rel="noopener nofollow ugc">Introduction to GitHub Actions - GitHub Docs</a></p>
<p>I’ve followed the instructions exactly, including copy-pasting the example code, but when I push and Github Actions runs it produces the following error:</p>
<pre><code class="lang-auto">Run npm install -g bats
npm WARN checkPermissions Missing write access to /usr/local/lib/node_modules
npm ERR! code EACCES
npm ERR! syscall access
npm ERR! path /usr/local/lib/node_modules
npm ERR! errno -13
npm ERR! Error: EACCES: permission denied, access '/usr/local/lib/node_modules'
npm ERR!  [Error: EACCES: permission denied, access '/usr/local/lib/node_modules'] {
npm ERR!   errno: -13,
npm ERR!   code: 'EACCES',
npm ERR!   syscall: 'access',
npm ERR!   path: '/usr/local/lib/node_modules'
npm ERR! }
npm ERR! 
npm ERR! The operation was rejected by your operating system.
npm ERR! It is likely you do not have the permissions to access this file as the current user
npm ERR! 
npm ERR! If you believe this might be a permissions issue, please double-check the
npm ERR! permissions of the file and its containing directories, or try running
npm ERR! the command again as root/Administrator.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/runner/.npm/_logs/2021-07-01T15_33_20_179Z-debug.log
Error: Process completed with exit code 243.
</code></pre>

**Comment by jsoref** [comment]
<p>You are correct, the instructions are flawed.</p>
<p>Reported at:</p><aside class="onebox githubissue" data-onebox-src="https://github.com/github/docs/issues/7866">
  <header class="source">

      <a href="https://github.com/github/docs/issues/7866" target="_blank" rel="noopener nofollow ugc">github.com/github/docs</a>
  </header>

  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Issue">
	  <svg width="60" height="60" class="github-icon" viewbox="0 0 14 16" aria-hidden="true"><path d="M7 2.3c3.14 0 5.7 2.56 5.7 5.7s-2.56 5.7-5.7 5.7A5.71 5.71 0 0 1 1.3 8c0-3.14 2.56-5.7 5.7-5.7zM7 1C3.14 1 0 4.14 0 8s3.14 7 7 7 7-3.14 7-7-3.14-7-7-7zm1 3H6v5h2V4zm0 6H6v2h2v-2z"></path></svg>
  </div>

  <div class="github-info-container">
    <h4>
      <a href="https://github.com/github/docs/issues/7866" target="_blank" rel="noopener nofollow ugc">Following Github Actions tutorial but getting an npm EACCES error</a>
    </h4>

    <div class="github-info">
      <div class="date">
        opened <span class="discourse-local-date" data-format="ll" data-date="2021-07-01" data-time="17:13:04" data-timezone="UTC">05:13PM - 01 Jul 21 UTC</span>
      </div>


      <div class="user">
        <a href="https://github.com/jsoref" target="_blank" rel="noopener nofollow ugc">
          <img alt="jsoref" src="https://user-images.githubusercontent.com/2119212/181101397-7f57755a-4844-4a0a-9eb5-29336f2511e9.png" class="onebox-avatar-inline" width="20" height="20">
          jsoref
        </a>
      </div>
    </div>

    <div class="labels">
        <span style="display:inline-block;margin-top:2px;background-color: #B8B8B8;padding: 2px;border-radius: 4px;color: #fff;margin-left: 3px;">
          content
        </span>
    </div>
  </div>
</div>

  <div class="github-row">
    <p class="github-body-container">&lt;!--
HUBBERS BEWARE! THE GITHUB/DOCS REPO IS PUBLIC TO THE ENTIRE INTERNET. OPE<span class="show-more-container"><a href="" rel="noopener" class="show-more">…</a></span><span class="excerpt hidden">N AN ISSUE IN GITHUB/DOCS-CONTENT INSTEAD.
--&gt;

&lt;!--
For questions, ask in Discussions: https://github.com/github/docs/discussions

Before you file an issue read the:
- Code of Conduct: https://github.com/github/docs/blob/main/CODE_OF_CONDUCT.md
- Contributing guide: https://github.com/github/docs/blob/main/CONTRIBUTING.md

Check to make sure someone hasn't already opened a similar issue: https://github.com/github/docs/issues
--&gt;

### What article on docs.github.com is affected?


https://docs.github.com/en/actions/learn-github-actions/introduction-to-github-actions#create-an-example-workflow

### What part(s) of the article would you like to see updated?


https://docs.github.com/en/actions/learn-github-actions/introduction-to-github-actions#create-an-example-workflow

### Additional information


https://github.com/jsoref/188088/runs/2965080849?check_suite_focus=true

I followed the instructions:
1. I created a repository https://github.com/jsoref/188088
2. I went to https://github.com/jsoref/188088/new/main?filename=.github%2Fworkflows%2Fblank.yml&amp;workflow_template=blank
3. I copy and pasted the filename and the contents from the help page
4. I clicked start commit
5. I pasted the url into the commit
6. I waited for the action to run and checked the results

It fails:
https://github.com/jsoref/188088/runs/2965080849?check_suite_focus=true

Precisely as reported here:
https://github.community/t/following-github-actions-tutorial-but-getting-an-npm-eacces-error/188088/1

Yes. I know how to do this right, but I should be able to follow the instructions (as @seddonr faithfully attempted) and get a working product.</span></p>
  </div>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>To learn one of many ways to fix this, you could read:</p><aside class="onebox allowlistedgeneric" data-onebox-src="https://danillolima.com/en/npm/how-to-fix-permissions-on-globally-installing-npm-packages-on-linux/">
  <header class="source">
      

      <a href="https://danillolima.com/en/npm/how-to-fix-permissions-on-globally-installing-npm-packages-on-linux/" target="_blank" rel="noopener nofollow ugc">danillo lima</a>
  </header>

  <article class="onebox-body">
    

<h3><a href="https://danillolima.com/en/npm/how-to-fix-permissions-on-globally-installing-npm-packages-on-linux/" target="_blank" rel="noopener nofollow ugc">How to fix permissions on globally installing npm packages on linux - danillo...</a></h3>

  <p>Is common have packages that works globally, they make the work more easy in some ways, provide functionalities, et al.. The difference in npm packages that are installed globally and locally is that you will setup a package like a program avaliable...</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>I’m sorry you hit a document that gave incorrect instructions.</p>

**Comment by seddonr** [comment]
<p>Thank you for reporting the problem.</p>
<p>I attempted to implement the suggested fix, <code>.github/workflows/learn-github-actions.yml</code> now looks like this:</p>
<pre><code class="lang-auto">name: learn-github-actions
on: [push]
jobs:
  check-bats-version:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-node@v2
      - run: mkdir ~/.npm-global
      - run: npm config set prefix '~/.npm-global'
      - run: echo 'export PATH=~/.npm-global/bin:$PATH' &gt;&gt; ~/.profile
      - run: source ~/.profile
      - run: npm install -g bats
      - run: bats -v
</code></pre>
<p>However now I get a different error:</p>
<pre><code class="lang-auto">Run bats -v
/home/runner/work/_temp/d705b0bc-8901-4f83-8cef-8b2c049df00e.sh: line 1: bats: command not found
Error: Process completed with exit code 127.
</code></pre>

**Comment by jsoref** [comment, marked as answer]
<h2>
<a name="problem-1" class="anchor" href="#problem-1"></a>Problem</h2>
<p>Each <code>run</code> statement almost certainly gets its own environment.</p>
<p>The one where it mattered the most was <code>source ~/.profile</code>, where you really needed it to last to the next step (<code>npm install ...</code>).</p>
<h2>
<a name="recommendation-2" class="anchor" href="#recommendation-2"></a>Recommendation</h2>
<p>I think you want this (untested):</p>
<pre><code class="lang-auto">name: learn-github-actions
on: [push]
jobs:
  check-bats-version:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-node@v2
      - run: |
           mkdir ~/.npm-global
           npm config set prefix '~/.npm-global'
           echo 'export PATH=~/.npm-global/bin:$PATH' &gt;&gt; ~/.profile
           source ~/.profile
           npm install -g bats
           bats -v
</code></pre>
<p>For reference, I picked the <code>|</code> based on:<br>
<a href="https://yaml-multiline.info/" class="onebox" target="_blank" rel="noopener nofollow ugc">https://yaml-multiline.info/</a><br>
I pretty much always consult this webpage…</p>
<p>Note: technically yo don’t need to use <code>~/.profile</code> here, you could just do:</p>
<pre><code class="lang-auto">name: learn-github-actions
on: [push]
jobs:
  check-bats-version:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-node@v2
      - run: |
           mkdir ~/.npm-global
           npm config set prefix '~/.npm-global'
           export PATH=~/.npm-global/bin:$PATH
           npm install -g bats
           bats -v
</code></pre>
<h2>
<a name="alternative-3" class="anchor" href="#alternative-3"></a>Alternative</h2>
<p>FWIW, there is an alternative,</p>
<p><a href="https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-an-environment-variable" class="onebox" target="_blank" rel="noopener nofollow ugc">https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-an-environment-variable</a><br>
<a href="https://docs.github.com/en/actions/reference/environment-variables" class="onebox" target="_blank" rel="noopener nofollow ugc">https://docs.github.com/en/actions/reference/environment-variables</a></p>
<p>However, in reading that pair of pages in context w/ a goal here, I expect that the average reader will be mislead. So, I guess I get to file another bug.</p>

**Comment by lar-mo** [comment]
<p>Thanks Jsoref:</p>
<p>Confirming this solution worked for me:</p>
<pre><code class="lang-auto">name: learn-github-actions
on: [push]
jobs:
  check-bats-version:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-node@v2
      - run: |
           mkdir ~/.npm-global
           npm config set prefix '~/.npm-global'
           export PATH=~/.npm-global/bin:$PATH
           npm install -g bats
           bats -v
</code></pre>
<p>Output from GitHub Job:</p>
<pre><code class="lang-auto">/home/runner/.npm-global/bin/bats -&gt; /home/runner/.npm-global/lib/node_modules/bats/bin/bats
+ bats@1.3.0
added 1 package from 1 contributor in 0.531s
Bats 1.3.0
</code></pre>
<p>Best~ lar-mo</p>
